cpp/src/arrow/csv/parser.cc

```diff
@@ -397,17 +397,24 @@ Status BlockParser::DoParseSpecialized(const char* start, uint32_t size, bool is
       return ParseError("Empty CSV file or block: cannot infer number of columns");
     }
   }
+  constexpr int32_t kMaxCols = 1024 * 1024;
```

Comment:
Is this limit useful? Is it just to guard against pathological cases or corrupt data?

Reply:
I removed it for now.  I think it is useful from the standpoint of we want to prevent causing OOMs in the standard library, but we can handle it as a separate JIRA since I don't think this is the appropriate place for the check.